### ISSUE TITLE
Memory leak in mautic:campaigns:trigger causing memory exhaustion on bigger campaigns

### ISSUE THREAD

**Issue by M4v3R** [mautic/mautic#6537]
**Please DO NOT report security vulnerabilities here. Send them to security@mautic.com instead.**

What type of report is this:

| Q  | A
| ---| ---
| Bug report? | Y
| Feature request? | N
| Enhancement? | N

## Description:
A memory leak exists somewhere in the code that runs during `mautic:campaigns:trigger` job. After each batch memory usage of PHP rises, until eventually a PHP memory exhaustion error is trigerred. The rate of memory leakage is dependant on the campaign event action type that is being executed. I noticed that the biggest leaks happen during "delete contact" action.

## If a bug:

| Q   | A
| --- | ---
| Mautic version | 2.14.0
| PHP version | 5.6.36

### Steps to reproduce:
1. Set a low `memory_limit` in PHP, for me 128M is enough to run Mautic while triggering the error quite early.
2. To monitor memory usage levels, in file `KickoffExecutioner.php`, after this line:
```php
$contacts = $this->kickoffContactFinder->getContacts($this->campaign->getId(), $this->limiter);
```
add:
```php
$this->output->writeln("\nMemory usage: " . memory_get_usage());
```
3. Import a few thousand test contacts to Mautic. You can use [this CSV](https://drive.google.com/open?id=12ab5tpWk8iT3alQEMwLrn5l0_P5AqbqW) which contains 5 thousand dummy contacts.
4. Create a segment that contains all contacts and run `mautic:segments:update`.
5. Create a campaign, select above segment and add a single action: "Delete contact"
6. Start the job from the command line: `php app/console mautic:campaigns:trigger`
7. Observe that the memory usage keeps rising, until eventualy PHP runs out of memory.
 
### Log errors: 

N/A

### More information:

My guess is that Doctrine's caching mechanism is the culprit here, although I've tried to clear this cache to no avail. I also tried to run `$this->campaignRepository->clear()` in `KickoffContactFinder.php` but this prevents the triggers from being executed correctly.

**Comment by npracht** [comment]
Hello @M4v3R ,

I'd advice you to test 2.14.1 version that includes https://github.com/mautic/mautic/pull/6224 and https://github.com/mautic/mautic/pull/6092 improves campaign performances.

**Comment by jbransen** [comment]
@npracht I am afraid this issue is not resolved, can we re-open this? With Mautic 2.14.2 I observe the same thing, based on the post above I added the following line for debugging:

```diff
+++ b/app/bundles/CampaignBundle/Executioner/ScheduledExecutioner.php
@@ -306,6 +306,7 @@ class ScheduledExecutioner implements ExecutionerInterface
     {
         $logs = $this->repo->getScheduled($eventId, $this->now, $this->limiter);
         while ($logs->count()) {
+            $this->output->writeln("\nMemory usage: " . memory_get_usage());
             try {
                 $this->scheduledContactFinder->hydrateContacts($logs);
             } catch (NoContactsFoundException $e) {
```

My `mautic:campaigns:trigger` output now is:
```
Triggering scheduled events
361966 total events(s) to be processed in batches of 100 contacts
      0/361966 [>---------------------------]   0%
Memory usage: 39559216
    100/361966 [>---------------------------]   0%
Memory usage: 50749760
    200/361966 [>---------------------------]   0%
Memory usage: 53260832
    300/361966 [>---------------------------]   0%
Memory usage: 55776896
    400/361966 [>---------------------------]   0%
[...SNIP...]
  29000/361966 [==>-------------------------]   8%
Memory usage: 797070104
  29100/361966 [==>-------------------------]   8%
Memory usage: 799628920
  29200/361966 [==>-------------------------]   8%
Memory usage: 802196280
```

I tried debugging, the problem is in the `scheduledContactFinder->hydrateContacts` call (or rather the `clear` call not freeing all memory), so I agree with M4v3R that it is most likely Doctrine cache, but I do not know how to solve this properly.

**Comment by jbransen** [comment]
@M4v3R See https://github.com/mautic/mautic/pull/6978
